### PR TITLE
Add signal when an item factory is added to the toolbar

### PR DESCRIPTION
**Pull request by brichet**
See https://github.com/jupyterlab/jupyterlab/issues/14366 for context.

## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/14366
The issue was first notified in https://github.com/jupyter/notebook/pull/6838

## Code changes

Adds a signal in `ToolbarWidgetRegistry`, which is emitted when an item factory is added to the registry.
This signal is caught by the `toolbarFactory` to render again the widget.

## User-facing changes

None

## Backwards-incompatible changes

None

EDIT:
Tested successfully on https://github.com/jupyter/notebook/pull/6838

### CODE REVIEW

File: packages/apputils/src/toolbar/factory.ts [hunk 1]
```diff
@@ -303,6 +303,25 @@ export function createToolbarFactory(
...
     };
 
+    const updateWidget = (
+      registry: IToolbarWidgetRegistry,
+      itemName: string
+    ) => {
+      const itemIndex = Array.from(items).findIndex(
+        item => item.name === itemName
+      );
+      if (itemIndex >= -1) {
```
Comment (fcollonval) [inline packages/apputils/src/toolbar/factory.ts:313]: I guess it is:

```suggestion
      if (itemIndex >= 0) {
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @brichet I think there is a small typo but otherwise it looks good.

File: packages/apputils/src/toolbar/factory.ts (hunk 1 above)
Comment (brichet) [reply packages/apputils/src/toolbar/factory.ts:313]: Thanks

**Comment by jtpio** [comment]
Thanks @brichet!

Wondering if there could be other situations than just toolbar items that could be affected by this new caching logic of the settings?

**Comment by brichet** [comment]
> Wondering if there could be other situations than just toolbar items that could be affected by this new caching logic of the settings?

This is hard to say, this is really an indirect effect, and there is probably no test on that behavior.
Actually, we were lucky it worked without trouble until now. It is likely that adding dependencies on other plugins with long loading times could have produced the same issue.

**Pull request merged**